packages/priority-queue/src/test/index.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * Internal dependencies
+ */
+import { createQueue } from '../';
+import requestIdleCallback from '../request-idle-callback';
+
+jest.mock( '../request-idle-callback', () => {
+	const emitter = new ( require.requireActual( 'events' ).EventEmitter )();
+
+	return Object.assign(
+		( callback ) => emitter.once( 'tick', () => callback( Date.now() ) ),
+		{ tick: () => emitter.emit( 'tick' ) },
+	);
+} );
+
+describe( 'createQueue', () => {
+	let queue;
+	beforeEach( () => {
+		queue = createQueue();
+	} );
+
+	describe( 'add', () => {
+		it( 'runs callback after processing waiting queue', () => {
+			const callback = jest.fn();
+
+			queue.add( {}, callback );
+
+			expect( callback ).not.toHaveBeenCalled();
+			requestIdleCallback.tick();
+			expect( callback ).toHaveBeenCalled();
+		} );
+
+		it( 'runs callbacks in order by distinct added element', () => {
+			const elementA = {};
+			const elementB = {};
+			const callbackElementA = jest.fn();
+			const callbackElementB = jest.fn();
+			queue.add( elementA, callbackElementA );
+			queue.add( elementB, callbackElementB );
+
+			expect( callbackElementA ).not.toHaveBeenCalled();
+			expect( callbackElementB ).not.toHaveBeenCalled();
+
+			// ElementA was added first, and should be called first after tick.
+			requestIdleCallback.tick();
+			expect( callbackElementA ).toHaveBeenCalledTimes( 1 );
+			expect( callbackElementB ).not.toHaveBeenCalled();
+
+			// ElementB will be be processed after second tick.
+			requestIdleCallback.tick();
+			expect( callbackElementA ).toHaveBeenCalledTimes( 1 );
+			expect( callbackElementB ).toHaveBeenCalledTimes( 1 );
+		} );
+
+		it( 'calls most recently added callback if added for same element', () => {
+			const element = {};
+			const callbackOne = jest.fn();
+			const callbackTwo = jest.fn();
+			queue.add( element, callbackOne );
+			queue.add( element, callbackTwo );
+
+			expect( callbackOne ).not.toHaveBeenCalled();
+			expect( callbackTwo ).not.toHaveBeenCalled();
+
+			requestIdleCallback.tick();
+			expect( callbackOne ).not.toHaveBeenCalled();
+			expect( callbackTwo ).toHaveBeenCalledTimes( 1 );
+		} );
+	} );
+
+	describe( 'flush', () => {
+		it( 'invokes all callbacks associated with element', () => {
+			const elementA = {};
+			const elementB = {};
+			const callbackElementA = jest.fn();
+			const callbackElementB = jest.fn();
+			queue.add( elementA, callbackElementA );
+			queue.add( elementB, callbackElementB );
+
+			expect( callbackElementA ).not.toHaveBeenCalled();
+			expect( callbackElementB ).not.toHaveBeenCalled();
+
+			queue.flush( elementA );
+
+			// Only ElementA callback should have been called (synchronously).
+			expect( callbackElementA ).toHaveBeenCalledTimes( 1 );
```

Comment:
It's kind of strange that we call this `flush`, but it doesn't flush a queue; it runs and removes a specific element from the queue.

Reply:
> It's kind of strange that we call this `flush`, but it doesn't flush a queue; it runs and removes a specific element from the queue.

Yeah, that's a fair concern. In a way I might think it could be explained in that each _context_ (element) from the outside behaves something like a queue of its own (or more like a single-entry stack) where the last-added callback is eventually invoked, but given how the waiting list is actually processed with a global "waiting list", it does seem to defy what one might expect from a flush here.

It makes me wonder whether there's even much value in having this key-by-`element` behavior, vs. just exposing the signature as `add( callback )`, and anywhere a distinct queue is needed, that implementation would simply create a new `createQueue()`. The only problem is that it creates a seperate scheduling for each queue, but I think this is something that would be reasonable to delegate to the default behavior of multiple `requestIdleCallback` (I expect they'd be processed in the order they're requested, for as long as time allows, in the same way the current implementation behaves).

Reply:
>In a way I might think it could be explained in that each context (element) from the outside behaves something like a queue of its own (or more like a single-entry stack) where the last-added callback is eventually invoked, but given how the waiting list is actually processed with a global "waiting list", it does seem to defy what one might expect from a flush here.

It's a queue of elements and a map of elements to callbacks.

>It makes me wonder whether there's even much value in having this key-by-element behavior, vs. just exposing the signature as add( callback ), and anywhere a distinct queue is needed, that implementation would simply create a new createQueue().

That makes more sense.

>The only problem is that it creates a seperate scheduling for each queue, but I think this is something that would be reasonable to delegate to the default behavior of multiple requestIdleCallback (I expect they'd be processed in the order they're requested, for as long as time allows, in the same way the current implementation behaves).

Yes, and it'd be easier to understand. This code wasn't entirely clear to me at first glance.